### DISCUSSION TITLE
CasePathable crashes on `get`

### DISCUSSION THREAD

**Discussion by Otbivnoe** [pointfreeco/swift-composable-architecture#3245] [Q&A] [answered]
I have a list with trips where each row has own reducer and state. For some reasons this line `ForEach(store.scope(state: \.[case: \.trips], action: \.trips))` causes the crash:

**Library version**: 1.11.2

<img width="663" alt="image" src="https://github.com/user-attachments/assets/80731d04-5779-4101-b05b-55450107935b">

My simplified code of the list:

```swift
struct TripsContent: Reducer {

    @ObservableState
    @CasePathable
    @dynamicMemberLookup
    enum State: Equatable {
        // other cases
        case trips(IdentifiedArrayOf<TripListItem.State>)
    }

    @CasePathable
    @dynamicMemberLookup
    enum Action {
        case trips(IdentifiedActionOf<TripListItem>)
    }

    var body: some Reducer<State, Action> {
        Reduce { state, action in
            return .none
        }
        .forEach(\.[case: \.trips], action: \.trips) {
            TripListItem()
        }
    }
}
```

```swift
struct TripsContentView: View {

    let store: StoreOf<TripsContent>

    var body: some View {
        switch store.state {
        case .trips:
            tripsList()
        }
    }

    private func tripsList() -> some View {
        ForEach(store.scope(state: \.[case: \.trips], action: \.trips)) { store in
            TripListItemView(store: store)
        }
    }
}
```

and a cell:

```swift 
struct TripListItem: Reducer {
    
    @ObservableState
    struct State: Identifiable, Equatable {
        var id: String { name }
        let name: String
    }

    @CasePathable
    enum Action {
        case onTap
    }

    var body: some Reducer<State, Action> {
        EmptyReducer()
    }
}
```

```swift
struct TripListItemView: View {

    let store: StoreOf<TripListItem>

    var body: some View {
        Button {
            store.send(.onTap)
        } label: {
            Text(store.name)
                .font(.system(size: 26))
        }
    }
}
```

What am I doing wrong? 🤔

**Comment by mbrandonw** [comment]
Hi @Otbivnoe, what version of Xcode/Swift are you using? Recent versions have improved their ability to do availability checking, and so should have prevented you from writing code that would get you into that `unavailable` getter. I would try Xcode 16 beta 3 to see if it now rejects this code.

**Reply by Otbivnoe** [reply]
Hey! I'm using Xcode 15.2 (Swift 5.9.2).

**Reply by Otbivnoe** [reply]
Installed the Version 15.4 and still no any helpful warnings... Can you help with the code change to get rid of this crash?

**Reply by ChloeMayhewELT** [reply, marked as answer]
Hi @Otbivnoe,

Rather than `\.[case: \.trips]`, you can just use `\.trips`.

**Reply by mbrandonw** [reply]
> Hey! I'm using Xcode 15.2 (Swift 5.9.2).

Ok, that would explain it. This code is not valid in Swift 6 becomes they fixed the availability checking:

<img width="967" alt="Screenshot 2024-07-18 at 3 09 23 PM" src="https://github.com/user-attachments/assets/4637c8e1-da99-4240-b28e-5b40b5dd6e69">

What you are attempting here is quite complex. An enum of identified arrays powering multiple `ForEach`s. For some complex, nested situations, we highly recommend making a property TCA domain for each layer.

So, you could have a `TripsList` feature to encapsulate the behavior of a list of trips, and it would be the one that uses the `.forEach` operator:

```swift
@Reducer
struct TripsList {
  @ObservableState
  struct State {
    var trips: IdentifiedArrayOf<TripListItem.State>
  }
  enum Action {
    case trips(IdentifiedActionOf<TripListItem>)
  }
  var body: some ReducerOf<Self> {
    Reduce<State, Action> { state, action in
        .none
    }
    .forEach(\.trips, action: \.trips) {
      TripListItem()
    }
  }
}
```

And then `TripsContent` would hold onto this domain in its enums, rather than the identified domain directly, and it would use the `ifCaseLet` operator to specify which reducer runs on which case:

```swift
struct TripsContent: Reducer {
  @ObservableState
  @CasePathable
  @dynamicMemberLookup
  enum State {
    // other cases
    case trips(TripsList.State)
  }

  @CasePathable
  enum Action {
    case trips(TripsList.Action)
  }

  var body: some Reducer<State, Action> {
    Reduce { state, action in
      return .none
    }
    .ifCaseLet(\.trips, action: \.trips) {
      TripsList()
    }
  }
}
```

And then the view would need to do something similar. It uses `if let` to unwrap the store in a particular case of the enum, and then `ForEach` to derive a store for each element in the collection:

```swift
@ViewBuilder
private func tripsList() -> some View {
  if let store = store.scope(state: \.trips, action: \.trips) {
    ForEach(store.scope(state: \.trips, action: \.trips)) { store in
      TripListItemView(store: store)
    }
  }
}
```

In general the tools work a lot better together when you just have a full feature at each level instead of taking shortcuts.

**Reply by Otbivnoe** [reply]
@mbrandonw Thanks for the detailed explanation!
